src/main/java/ch/ethz/seb/sebserver/webservice/servicelayer/lms/impl/olat/OlatLmsRestTemplate.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2021 ETH Zürich, Educational Development and Technology (LET)
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package ch.ethz.seb.sebserver.webservice.servicelayer.lms.impl.olat;
+
+import java.io.IOException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.web.client.RestTemplate;
+import org.springframework.security.oauth2.client.token.grant.client.ClientCredentialsResourceDetails;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.HttpRequest;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.client.ClientHttpRequestExecution;
+import org.springframework.http.client.ClientHttpRequestInterceptor;
+import org.springframework.http.client.ClientHttpResponse;
+
+public class OlatLmsRestTemplate extends RestTemplate {
+
+    private static final Logger log = LoggerFactory.getLogger(OlatLmsRestTemplate.class);
+
+    private String token;
+    private ClientCredentialsResourceDetails details;
+
+    public OlatLmsRestTemplate(ClientCredentialsResourceDetails details) {
+        super();
+        this.details = details;
+
+        // Add X-OLAT-TOKEN request header to every request done using this RestTemplate
+        this.getInterceptors().add(new ClientHttpRequestInterceptor(){
+            @Override
+            public ClientHttpResponse intercept(HttpRequest request, byte[] body, ClientHttpRequestExecution execution) throws IOException {
+                // if there's no token, authenticate first
+                if (token == null) { authenticate(); }
+                // when authenticating, just do a normal call
+                else if (token == "authenticating") { return execution.execute(request, body); }

```

Comment:
Use **"authenticating".equals(token)** here instead of == because in Java == checks on object level, if the object on the right side is the same as the object on the left side. Since there is always a new String object created with "authenticating" this will not work as expected.

Alternatively you can define "authenticate" as a constant: **private static final AUTH_MARK = "authenticate"**. If you now always use the constant, you can use the == operator instead of equals.